pkg/export/disk/disk.go

```diff
@@ -0,0 +1,271 @@
+	conn, ok := r.openConnections[connectionName]
+	if !ok {
+		return fmt.Errorf("connection not found: %s for disk driver", connectionName)
+	}
+	err := os.RemoveAll(conn.Path)
+	delete(r.openConnections, connectionName)
+	return err
+}
+
+func (r *Writer) Publish(_ context.Context, connectionName string, data interface{}, topic string) error {
+	conn, ok := r.openConnections[connectionName]
+	if !ok {
+		return fmt.Errorf("connection not found: %s for disk driver", connectionName)
+	}
+
+	var violation util.ExportMsg
+	if violation, ok = data.(util.ExportMsg); !ok {
+		return fmt.Errorf("invalid data type, cannot convert data to exportMsg")
+	}
+
+	if violation.Message == "audit is started" {
+		err := conn.handleAuditStart(violation.ID, topic)
+		if err != nil {
+			return fmt.Errorf("error handling audit start: %w", err)
+		}
+		r.openConnections[connectionName] = conn
+	}
+
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("error marshaling data: %w", err)
+	}
+
+	if conn.File == nil {
+		return fmt.Errorf("no file to write the violation in")
+	}
+
+	_, err = conn.File.WriteString(string(jsonData) + "\n")
+	if err != nil {
+		return fmt.Errorf("error writing message to disk: %w", err)
+	}
+
+	if violation.Message == "audit is completed" {
+		err := conn.handleAuditEnd(topic)
+		if err != nil {
+			return fmt.Errorf("error handling audit end: %w", err)
+		}
+		conn.File = nil
+		conn.currentAuditRun = ""
+		r.openConnections[connectionName] = conn
+	}
+	return nil
+}
+
+func (conn *Connection) handleAuditStart(auditID string, topic string) error {
+	conn.currentAuditRun = strings.ReplaceAll(auditID, ":", "_")
+
+	// Ensure the directory exists
+	dir := path.Join(conn.Path, topic)
+	if err := os.MkdirAll(dir, 0o755); err != nil {
+		return fmt.Errorf("failed to create directories: %w", err)
+	}
+
+	file, err := os.OpenFile(path.Join(dir, appendExtension(conn.currentAuditRun, "txt")), os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+	conn.File = file
+	err = retry.OnError(retry.DefaultBackoff, func(_ error) bool {
+		return true
+	}, func() error {
+		return syscall.Flock(int(conn.File.Fd()), syscall.LOCK_EX)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to acquire lock: %w", err)
+	}
+	return nil
+}
+
+func (conn *Connection) handleAuditEnd(topic string) error {
+	if err := retry.OnError(retry.DefaultBackoff, func(_ error) bool {
+		return true
+	}, conn.unlockAndCloseFile); err != nil {
+		return fmt.Errorf("error closing file: %w, %s", err, conn.currentAuditRun)
+	}
+	conn.File = nil
+
+	readyFilePath := path.Join(conn.Path, topic, appendExtension(conn.currentAuditRun, "log"))
+	if err := os.Rename(path.Join(conn.Path, topic, appendExtension(conn.currentAuditRun, "txt")), readyFilePath); err != nil {
+		return fmt.Errorf("failed to rename file: %w, %s", err, conn.currentAuditRun)
+	}
```

Comment:
suggestion: Consider logging when we remove or rename files to help with debugging any future coordination issues between reader and writer